sbin/iscontrol/iscontrol.8
.\" Copyright (c) 2007-2010 Daniel Braniss <danny@cs.huji.ac.il>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: head/sbin/iscontrol/iscontrol.8 255570 2013-09-14 15:29:06Z trasz $
.\"
.Dd August 9, 2010
.Dt ISCONTROL 8
.Os
.Sh NAME
.Nm iscontrol
.Nd login/negotiator/control for an iSCSI initiator session
.Sh SYNOPSIS
.Nm
.Op Fl dv
.Oo
.Fl c Ar file
.Op Fl n Ar nickname
.Oc
.Op Fl p Ar pidfile
.Op Fl t Ar target
.Op Ar variable Ns = Ns Ar value
.Sh DESCRIPTION
Internet SCSI (iSCSI) is a network protocol standard, that allows the
use of the SCSI protocol over TCP/IP networks,
the
.Nm
program is the userland side of an iSCSI session, see
.Xr iscsi_initiator 4 .
It has 2 modes of operation, if -d (discovery session) is specified,
it will print out the
.Em target names
returned by the target and exit.
In the second mode, it will, after a successful login/negotiation, run
in daemon mode, monitoring the connection, and will try to reconnect
in case of a network/target failure.
It will terminate/logout the session
when a SIGHUP signal is received.
The flags are as follows:
.Bl -tag -width variable=value
.It Fl c Ar file
a file containing configuration
.Em key-options ,
see
.Xr iscsi.conf 5 .
.It Fl d
do a
.Em discovery session
and exit.
.It Fl n Ar nickname
if
.Sy -c file
is specified, then search for the block named
.Em nickname
in that file, see
.Xr iscsi.conf 5 .
.It Fl p Ar pidfile
will write the process ID of the session to the specified
.Em pidfile
.It Fl t Ar target
the target's IP address or name.
.It Fl v
verbose mode.
.It Ar variable Ns = Ns Ar value
see
.Xr iscsi.conf 5
for the complete list of variables/options and their
possible values.
.El
.Sh EXAMPLES
.Dl iscontrol -dt myiscsitarget
.Pp
will start a
.Em discovery session
with the target and
print to stdout the list of available targetnames/targetadresses.
Note: this listing does not necessarily mean availability, since
depending on the target configuration, a discovery session might
not need login/access permission, but a
.Em full session
certainly does.
.sp
.Dl iscontrol -c /etc/iscsi.conf -n myiscsi
.Pp
will read options from
.Pa /etc/iscsi.conf ,
use the targetaddress
found in the block nicknamed myiscsi, login and negotiate
whatever options are specified, and start an iscsi-session.
.Sh SEE ALSO
.Xr da 4 ,
.Xr iscsi_initiator 4 ,
.Xr sa 4 ,
.Xr iscsi.conf 5 ,
.Xr iscsictl 8 ,
.Xr camcontrol 8
.Sh STANDARDS
RFC 3720
.\"Sh HISTORY
.Sh BUGS
.Nm
should probably load the iscsi_initiator module if needed.
.br
Not all functions/specifications have been implemented yet, noticeably
missing are the Task Management Functions.
The error recovery, though not
.Em fully compliant
does a brave effort to recover from network disconnects.


sbin/iscontrol/iscontrol.c
/*-
 * Copyright (c) 2005-2010 Daniel Braniss <danny@cs.huji.ac.il>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */
/*
 | $Id: iscontrol.c,v 2.2 2006/12/01 09:11:56 danny Exp danny $
 */
/*
 | the user level initiator (client)
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: head/sbin/iscontrol/iscontrol.c 254657 2013-08-22 14:02:34Z trasz $");

#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/sysctl.h>

#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <sys/ioctl.h>
#include <netdb.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <camlib.h>

#include <dev/iscsi_initiator/iscsi.h>
#include "iscontrol.h"

static char version[] = "2.3.1"; // keep in sync with iscsi_initiator

#define USAGE "[-v] [-d] [-c config] [-n name] [-t target] [-p pidfile]"
#define OPTIONS	"vdc:t:n:p:"

token_t AuthMethods[] = {
     {"None",	NONE},
     {"KRB5",	KRB5},
     {"SPKM1",	SPKM1},
     {"SPKM2",	SPKM2},
     {"SRP",	SRP},
     {"CHAP",	CHAP},
     {0, 0}
};

token_t	DigestMethods[] = {
     {"None",	0},
     {"CRC32",	1},
     {"CRC32C",	1},
     {0, 0}
};

u_char	isid[6 + 6];
/*
 | Default values
 */
isc_opt_t opvals = {
     .port			= 3260,
     .sockbufsize		= 128,
     .iqn			= "iqn.2005-01.il.ac.huji.cs:",

     .sessionType		= "Normal",
     .targetAddress		= 0,
     .targetName		= 0,
     .initiatorName		= 0,
     .authMethod		= "None",
     .headerDigest		= "None,CRC32C",
     .dataDigest		= "None,CRC32C",
     .maxConnections		= 1,
     .maxRecvDataSegmentLength	= 64 * 1024,
     .maxXmitDataSegmentLength	= 8 * 1024, // 64 * 1024,
     .maxBurstLength		= 128 * 1024,
     .firstBurstLength		= 64 * 1024, // must be less than maxBurstLength
     .defaultTime2Wait		= 0,
     .defaultTime2Retain	= 0,
     .maxOutstandingR2T		= 1,
     .errorRecoveryLevel	= 0,

     .dataPDUInOrder		= TRUE,
     .dataSequenceInOrder	= TRUE,

     .initialR2T		= TRUE,
     .immediateData		= TRUE,
};

static void
usage(const char *pname)
{
	fprintf(stderr, "usage: %s " USAGE "\n", pname);
	exit(1);
}

int
lookup(token_t *tbl, char *m)
{
     token_t	*tp;

     for(tp = tbl; tp->name != NULL; tp++)
	  if(strcasecmp(tp->name, m) == 0)
	       return tp->val;
     return 0;
}

int
main(int cc, char **vv)
{
     int	ch, disco;
     char	*pname, *pidfile, *p, *q, *ta, *kw, *v;
     isc_opt_t	*op;
     FILE	*fd;
     size_t	n;

     op = &opvals;
     iscsidev = "/dev/"ISCSIDEV;
     fd = NULL;
     pname = vv[0];
     if ((pname = basename(pname)) == NULL)
	  err(1, "basename");

     kw = ta = 0;
     disco = 0;
     pidfile = NULL;
     /*
      | check for driver & controller version match
      */
     n = 0;
#define VERSION_OID_S	"net.iscsi_initiator.driver_version"
     if (sysctlbyname(VERSION_OID_S, 0, &n, 0, 0) != 0) {
	  if (errno == ENOENT)
		errx(1, "sysctlbyname(\"" VERSION_OID_S "\") "
			"failed; is the iscsi driver loaded?");
	  err(1, "sysctlbyname(\"" VERSION_OID_S "\")");
     }
     v = malloc(n+1);
     if (v == NULL)
	  err(1, "malloc");
     if (sysctlbyname(VERSION_OID_S, v, &n, 0, 0) != 0)
	  err(1, "sysctlbyname");

     if (strncmp(version, v, 3) != 0)
	  errx(1, "versions mismatch");

     while((ch = getopt(cc, vv, OPTIONS)) != -1) {
	  switch(ch) {
	  case 'v':
	       vflag++;
	       break;
	  case 'c':
	       fd = fopen(optarg, "r");
	       if (fd == NULL)
		    err(1, "fopen(\"%s\")", optarg);
	       break;
	  case 'd':
	       disco = 1;
	       break;
	  case 't':
	       ta = optarg;
	       break;
	  case 'n':
	       kw = optarg;
	       break;
	  case 'p':
	       pidfile = optarg;
	       break;
	  default:
	       usage(pname);
	  }
     }
     if(fd == NULL)
	  fd = fopen("/etc/iscsi.conf", "r");

     if(fd != NULL) {
	  parseConfig(fd, kw, op);
	  fclose(fd);
     }
     cc -= optind;
     vv += optind;
     if(cc > 0) {
	  if(vflag)
	       printf("adding '%s'\n", *vv);
	  parseArgs(cc, vv, op);
     }
     if(ta)
	  op->targetAddress = ta;

     if(op->targetAddress == NULL) {
	  warnx("no target specified!");
	  usage(pname);
     }
     q = op->targetAddress;
     if(*q == '[' && (q = strchr(q, ']')) != NULL) {
	  *q++ = '\0';
	  op->targetAddress++;
     } else
	  q = op->targetAddress;
     if((p = strchr(q, ':')) != NULL) {
	  *p++ = 0;
	  op->port = atoi(p);
	  p = strchr(p, ',');
     }
     if(p || ((p = strchr(q, ',')) != NULL)) {
	  *p++ = 0;
	  op->targetPortalGroupTag = atoi(p);
     }
     if(op->initiatorName == 0) {
	  char	hostname[MAXHOSTNAMELEN];

	  if(op->iqn) {
	       if(gethostname(hostname, sizeof(hostname)) == 0)
		    asprintf(&op->initiatorName, "%s:%s", op->iqn, hostname);
	       else
		    asprintf(&op->initiatorName, "%s:%d", op->iqn, (int)time(0) & 0xff); // XXX:
	  }
	  else {
	       if(gethostname(hostname, sizeof(hostname)) == 0)
		    asprintf(&op->initiatorName, "%s", hostname);
	       else
		    asprintf(&op->initiatorName, "%d", (int)time(0) & 0xff); // XXX:
	  }
     }
     if(disco) {
	  op->sessionType = "Discovery";
	  op->targetName = 0;
     }
     op->pidfile = pidfile;
     fsm(op);

     exit(0);
}
